### ISSUE TITLE
Cached images not being displayed and Reload Image

### ISSUE THREAD

**Issue by beastie1** [OtterBrowser/otter-browser#1149]
I don't know if these are 2 unrelated issues.

http://im9.eu/album/otter-browser

**Comment by Emdek** [comment]
@beastie1, please elaborate more, this is too laconic. :-P

**Comment by beastie1** [comment]
Yes, sorry I accidentally pressed Enter when I was editing the title hehe!

**Comment by beastie1** [comment]
(Edited the first comment)

**Comment by Emdek** [comment]
@beastie1, currently there is no way to pick specific images to get exception from this rule.
Perhaps temporarily disabling that mode while reloading would help, but probably it would also affect other images on the same page. I'll consult that with @annulen.

The second part is odd, they seem to indeed not to be saved in cache at all.
But since we don't mess with cache for these images it seem to be upstream bug / feature.

**Comment by Emdek** [comment]
@beastie1, I've tested both QtWebKit legacy and NG and QtWebEngine, all behave the same way, directly viewed images seem to not be cached (at least not on the disk).
Of course there is chance that something wrong is on our side, but according to my earlier tests it works just fine.

I've also confirmed that simply enabling images for a moment is no go, it loads all of them then, without invoking image reload.
But there is a chance, I could store URL of reloaded image and then filter out all others, that could work.

**Comment by beastie1** [comment]
It sounds good. Let's hope such hack doesn't affect performance too much.

Thanks for looking into it.

**Comment by Emdek** [comment]
@beastie1, nah, there should be no difference, it would do one more check if mode is set to cached images only and list of exceptions is not empty.

**Comment by Emdek** [comment]
@beastie1, above change should help, although as side effect it draws image placehoders for all other images...
But there is also a bonus, it allows to also override content blocking rules (for both images and frames).

**Comment by beastie1** [comment]
Great! It also fixes what I reported in #820 about favicons not being displayed when images are disabled globally.

I still can notice 3 issues:
* Placeholders are only visible in "No images" mode, but not in "Cached images"
* On some websites (e.g. Google Images) images are still displayed even if they are disabled both globally and in the current tab
* When in "No images" mode, opening images using any method other than "Reload Image" results in an "Error 301" (e.g. opening a **<a href="image.png">** link, right-clicking the placeholder and opening in a new/background tab, pasting an image URL in the address bar)

**Comment by Emdek** [comment]
@beastie1, as already mentioned, we don't have control over placeholders. And yes, this inconsistency is annoying.
Exceptions might be caused by either embedding images or using some format that is not listed (Google might push WebP to save traffic), we will need to wait for some improvements in QtWebKit, like built-in flag to disable images.
I guess that we should add exception for viewing images directly, although in theory no images should mean no images. ;-)

**Comment by Emdek** [comment]
Ability to disable all images in consistent way landed upstream, but it will be safer to wait until next release to start using it:
https://github.com/annulen/webkit/commit/a681d7731ea12e4b4204e64f0b540b06d5f2fbd2

Sadly I haven't found a sane way to control placeholders...
Perhaps we could consider two modes, defaulting to proper blocking by WebKit itself and another, which blocks images old way, thus producing effect of "missing image".
